lang/cs/Org.Apache.REEF.IMRU.Examples/PipelinedBroadcastReduce/PipelinedBroadcastAndReduceWithRandomFailures.cs

```diff
@@ -0,0 +1,389 @@
+using Org.Apache.REEF.Common.Tasks;
+using Org.Apache.REEF.IMRU.API;
+using Org.Apache.REEF.IMRU.OnREEF.IMRUTasks;
+using Org.Apache.REEF.IMRU.OnREEF.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Implementations.Configuration;
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+using Org.Apache.REEF.Utilities.Logging;
+using System.Threading;
+
+namespace Org.Apache.REEF.IMRU.Examples.PipelinedBroadcastReduce
+{
+    /// <summary>
+    /// IMRU program that performs broadcast and reduce with fault tolerance.
+    /// </summary>
+    public sealed class PipelinedBroadcastAndReduceWithRandomFailures : PipelinedBroadcastAndReduce
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(PipelinedBroadcastAndReduceWithRandomFailures));
+
+        [Inject]
+        private PipelinedBroadcastAndReduceWithRandomFailures(IIMRUClient imruClient) : base(imruClient)
+        {
+        }
+
+        /// <summary>
+        /// Runs the actual broadcast and reduce job with fault tolerance
+        /// </summary>
+        internal void Run(int numberofMappers, int chunkSize, int numIterations, int dim, int mapperMemory, int updateTaskMemory, 
+            int maxRetryNumberInRecovery, float failProb, int minTimeoutInSec, int expThroughput)
+        {
+            var results = _imruClient.Submit<int[], int[], int[], Stream>(
+                CreateJobDefinitionBuilder(numberofMappers, chunkSize, numIterations, dim, mapperMemory, updateTaskMemory)
+                    .SetMapFunctionConfiguration(BuildMapperFunctionConfig(dim, failProb, minTimeoutInSec, expThroughput))
+                    .SetMaxRetryNumberInRecovery(maxRetryNumberInRecovery)
+                    .Build());
+        }
+
+        protected override IMRUJobDefinitionBuilder CreateJobDefinitionBuilder(int numberofMappers, int chunkSize, int numIterations, int dim, int mapperMemory, int updateTaskMemory)
+        {
+            return new IMRUJobDefinitionBuilder()
+                    .SetUpdateTaskStateConfiguration(UpdateTaskStateConfiguration())
+                    .SetMapTaskStateConfiguration(MapTaskStateConfiguration())
+                    .SetUpdateFunctionConfiguration(UpdateFunctionConfig(numberofMappers, numIterations, dim))
+                    .SetMapInputCodecConfiguration(MapInputCodecConfiguration())
+                    .SetUpdateFunctionCodecsConfiguration(UpdateFunctionCodecsConfiguration())
+                    .SetReduceFunctionConfiguration(ReduceFunctionConfiguration())
+                    .SetMapInputPipelineDataConverterConfiguration(MapInputDataConverterConfig(chunkSize))
+                    .SetMapOutputPipelineDataConverterConfiguration(MapOutputDataConverterConfig(chunkSize))
+                    .SetPartitionedDatasetConfiguration(PartitionedDatasetConfiguration(numberofMappers))
+                    .SetJobName("BroadcastReduce")
```

Comment:
It is probably best to add `WithRandomFailure` here? That way, we see in the YARN app list what is going on.